File: Data Structures & Algorithms/course-schedule/submission-1.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0; i<numCourses;i++){
            graph.add(new ArrayList<>());
        }

        for(int[] pre: prerequisites){
            int course = pre[0];
            int prereq = pre[1];
            graph.get(course).add(prereq);
        }
        int[] state = new int[numCourses];

        for(int i=0; i<numCourses;i++){
            if(hasCycle(i, graph, state)){
                return false;
            }
        }
        return true;
    }
    private boolean hasCycle(int course, List<List<Integer>> graph, int[] state){
        if(state[course] == 1){
            return true;
        }
        if(state[course] == 2){
            return false;
        }
        state[course] = 1; 
        for(int prereq: graph.get(course)){
            if(hasCycle(prereq, graph, state)){
                return true;
            }
        }
        state[course] = 2;
        return false;
    }
}

File: Data Structures & Algorithms/course-schedule-ii/submission-1.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        
        adj = [[] for c in range(numCourses)]

        indegree = [0] * numCourses

        for course, prereq in prerequisites:
            adj[prereq].append(course)
            indegree[course] += 1 
        
        q = deque()

        for course in range(numCourses):
            if indegree[course] == 0:
                q.append(course)

        order = []

        while q:
            course = q.popleft()
            order.append(course)

            for next_course in adj[course]:
                indegree[next_course] -= 1

                if indegree[next_course] == 0:
                    q.append(next_course)
        
        return order if len(order) == numCourses else []